### PR TITLE
Track clicks when element inside of <a href> tag  is clicked.

### PR DESCRIPTION
**Pull request by fleitz**
<code>
&lt;a href="#foo">
&lt;div>
&lt;img src="clickable_image">
&lt;/div>
&lt;/a>
</code>
<br>
Expected result:
Clicking on image results in tracking

Actual Result:
Clicking on image does not result in tracking.

### CODE REVIEW

**Comment by ankane** [comment]
Hey @fleitz, thanks for the PR, and sorry for the delay (have been focused on other projects). This change makes sense to me. Instead of setting `matchedElement` on the event, it would be good to make `this` the matched element (as libraries like jQuery do).

**Comment by fleitz** [comment]
Yeah, that sounds like a way cleaner way to implement.

File: src/index.js [hunk 1]
```diff
@@ -252,14 +258,13 @@ function cleanObject(obj) {
   return obj;
 }
 
-function eventProperties(e) {
-  let target = e.target;
```
Comment (ankane) [inline src/index.js:256]: Let's use `this` for the `target` and keep the rest of the method intact, so #55 will work.

File: src/index.js [hunk 2]
```diff
@@ -100,17 +100,23 @@ function matchesSelector(element, selector) {
     element.webkitMatchesSelector;
 
   if (matches) {
-    return matches.apply(element, [selector]);
+    if(matches.apply(element, [selector])){
```
Comment (ankane) [inline src/index.js:103]: Small one: please use the existing style for spacing.

File: src/index.js [hunk 3]
```diff
@@ -418,25 +423,31 @@ ahoy.trackView = function (additionalProperties) {
...
+  // onEvent will callback with this = a#text and e.target = p#text
   onEvent("click", "a, button, input[type=submit]", function (e) {
-    let target = e.target;
-    let properties = eventProperties(e);
+    let target = this;
+    let properties = elementProperties(this); 
     properties.text = properties.tag == "input" ? target.value : (target.textContent || target.innerText || target.innerHTML).replace(/[\s\r\n]+/g, " ").trim();
     properties.href = target.href;
+    if(e.target != this){
+      properties.clickedElement = elementProperties(e.target);
```
Comment (ankane) [inline src/index.js:435]: Let's avoid extra properties (users will be able to add them with #55 if needed).

**Comment by fleitz** [comment]
@ankane implemented.

**Review by ankane — COMMENTED** [review]
Thanks @fleitz, looking better. Added a few comments inline.

File: src/index.js (hunk 1 above)
Comment (fleitz) [reply src/index.js:256]: oh I got rid of eventProperties and changed it to elementProperties as all it did was take the target from the event.

Comment (fleitz) [reply src/index.js:256]: Changed it back to eventProperties and made it clear what 'this' is.

**Review by ankane — COMMENTED** [review]
Thanks @fleitz, added a few more comments. Please follow the existing spacing everywhere.

File: src/index.js [hunk 4]
```diff
@@ -100,17 +100,23 @@ function matchesSelector(element, selector) {
...
-    return false;
+    return null;
   }
 }
 
 function onEvent(eventName, selector, callback) {
   document.addEventListener(eventName, function (e) {
-    if (matchesSelector(e.target, selector)) {
-      callback(e);
+    var matchedElement = matchesSelector(e.target, selector);
```
Comment (ankane) [inline src/index.js:117]: `var` -> `let`

File: src/index.js [hunk 5]
```diff
@@ -418,9 +424,12 @@ ahoy.trackView = function (additionalProperties) {
...
 
 ahoy.trackClicks = function () {
+  // Example: <a id="link" href="/foo"><p id="text">Foo</p></a>
+  // e.target on Chrome (and likely every other browser) is not a#link tag but p#text 
+  // onEvent will callback with this = a#text and e.target = p#text
   onEvent("click", "a, button, input[type=submit]", function (e) {
-    let target = e.target;
-    let properties = eventProperties(e);
+    let target = this;
+    let properties = eventProperties.call(target,e);
```
[the remark this answers is not in the document]
Comment (fleitz) [reply src/index.js]: fixed everywhere, removed spaces

File: src/index.js [hunk 6]
```diff
@@ -100,17 +100,23 @@ function matchesSelector(element, selector) {
     element.webkitMatchesSelector;
 
   if (matches) {
-    return matches.apply(element, [selector]);
+    if (matches.apply(element, [selector])) {
+      return element;
+    } else if(element.parentElement) {
+      return matchesSelector(element.parentElement,selector)
```
[the remark this answers is not in the document]
Comment (fleitz) [reply src/index.js]: fixed spacing

File: src/index.js (hunk 4 above)
Comment (fleitz) [reply src/index.js]: changed var to let

**Comment by ankane** [comment]
Thanks @fleitz! This is a great feature.

**Pull request merged**